src/lib/logical_query_plan/abstract_lqp_node.hpp

```diff
@@ -334,4 +377,67 @@ class AbstractLQPNode : public std::enable_shared_from_this<AbstractLQPNode>, pr
   // @}
 };
 
+/**
+ * LQP node types should derive from this, instead of directly from AbstractLQPNode, in order to enable the
+ * <NodeType>::make() that allows for a clean notation when building LQPs via code by allowing to pass in a nodes
+ * child(ren) as the last argument(s).
+ *
+ * const auto input_lqp =
+ * PredicateNode::make(_mock_node_a, PredicateCondition::Equals, 42,
+ *   PredicateNode::make(_mock_node_b, PredicateCondition::GreaterThan, 50,
+ *     PredicateNode::make(_mock_node_b, PredicateCondition::GreaterThan, 40,
+ *       ProjectionNode::make_pass_through(
+ *         PredicateNode::make(_mock_node_a, PredicateCondition::GreaterThanEquals, 90,
+ *           PredicateNode::make(_mock_node_c, PredicateCondition::LessThan, 500,
+ *             _mock_node))))));
+ */
+template <typename DerivedNode>
+class AbstractLQPNodeEnableMake : public AbstractLQPNode {
+ public:
+  using AbstractLQPNode::AbstractLQPNode;
+
+  using Base = AbstractLQPNodeEnableMake<DerivedNode>;
+
+  template <int N, typename... Ts>
+  using NthTypeOf = typename std::tuple_element<N, std::tuple<Ts...>>::type;
+
+  template <typename... Args>
+  static std::shared_ptr<DerivedNode> make(Args... args) {
+    // clang-format off
+    if constexpr(sizeof...(Args) > 0) {
+      if constexpr(std::is_convertible_v<NthTypeOf<sizeof...(Args)-1, Args...>, std::shared_ptr<AbstractLQPNode>>) {
```

Comment:
Any reason you split this into two ifs?

Reply:
yes, there was some reason... it seems like constexpr if will evaluate both sides of a &&, even if the left side is false.

Reply:
If this is true, doesn't this apply to the check some lines below as well?

Reply:
(if constexpr *does* short-circuit evaluation, but the predicates have to be valid. `NthTypeOf<sizeof...(Args)-1, Args...>` is not.)

Reply:
fixed and added comment